CMakeLists.txt

```diff
@@ -167,12 +167,10 @@ else()
   target_link_libraries(rmm PUBLIC CUDA::cudart)
 endif(CUDA_STATIC_RUNTIME)
 
-target_link_libraries(rmm PUBLIC CUDA::cuda_driver)
-
 if(USE_NVTX)
   message(STATUS "Using Nvidia Tools Extension")
-  target_link_libraries(rmm PUBLIC CUDA::nvToolsExt)
```

Comment:
Why remove this? When we add NVTX annotations we will just have to re-add it.

Reply:
We use the header only version of NVTX. We haven't needed to link against this for ages.

Reply:
> * [x]  Correctly set `NVTX_DISABLE` when `USE_NVTX` is off

I think it's referring to this in the top of the PR. The CMake target is the shared library only.

Reply:
I'll have to take a closer look at this. But generally speaking, it's worth pointing out that `target_link_libraries` has its use with header-only libraries, even though in that case no actually library exists. It basically just expresses that a target depends on another one, and that might just properly set include paths so that the headers can be found, or might set compiler flags or whatever.

In this particular case, I know that an actual `libnvToolsExt.so` exists, and so I suppose it does get linked, but I'm not sure there's any harm in that, in particular because this does make sure that the header for it is found (which it probably would be anyway, because other stuff from the toolkit is used, but it still fundamentally makes sense to depend on the things you're using.)

Reply:
RMM is going towards being header only, there wouldn't be anything to link `libnvToolsExt.so` against. I doubt there is a cmake target for the nvtx header only version. If we want one, we should be adding it to the NVTX C++ wrappers: https://github.com/NVIDIA/NVTX/tree/dev/cpp

Reply:
A header-only library can still depend on actual libraries (like, e.g., `libucdart` in the case of rmm). Of course the header-only library itself can't link to it, but a consumer still needs to.

So first of all, I agree that since rmm isn't using nvtx at all right now, fundamentally there shouldn't be anything nvtx related in its build, either. But since it sounds like it'll be added soon, of course it doesn't make too much sense to undo it now, either.

I do not think it's a good idea to set `NVTX_DISABLE` in rmm. That means a downstream consumer will automatically get this flag set, too, and they may be surprised that by means of using rmm their profile markers disappear.

Also, is there a link to read about header-only nvtx? It looks to me like CUDA 10.2 still has the regular nvtx, is that not going to be supported by rmm? I think you're right, though, that the `CUDA::nvToolsExt` target is for the conventional library.

Reply:
> Also, is there a link to read about header-only nvtx? It looks to me like CUDA 10.2 still has the regular nvtx, is that not going to be supported by rmm? I think you're right, though, that the CUDA::nvToolsExt target is for the conventional library.

Not that I'm aware of. The toolkit ships with both the shared library and header only versions. The header only version is in `cuda/include/nvtx3/`.

Reply:
> The intention was only to disable it for librmm.so. This becomes irrelevant when RMM is header only.

@germasch explained why it's relevant for a header-only library. It's a cmake thing -- "That means a downstream consumer will automatically get this flag set, too, and they may be surprised that by means of using rmm their profile markers disappear."

Header only NVTX3 source is here (dev branch, not released yet): https://github.com/NVIDIA/NVTX/tree/dev

Reply:
> @germasch explained why it's relevant for a header-only library. It's a cmake thing -- "That means a downstream consumer will automatically get this flag set, too, and they may be surprised that by means of using rmm their profile markers disappear."

It's not relevant because the option will be removed. There's no point in setting it when RMM isn't building it's own artifact anymore.

Furthermore, even while librmm is being built, `-DDISABLE_NVTX` should only be present in building `librmm.so`. If that's leaking to downstream projects, then that is a bug. 

> Header only NVTX3 source is here (dev branch, not released yet): https://github.com/NVIDIA/NVTX/tree/dev

The `NVTX3` C++ wrappers are distinct from the nvtx3 C API. The header only, NVTX C API has been shipping with the CUDA toolkit at least since CUDA 10. It's the same code that is present here: https://github.com/NVIDIA/NVTX/tree/release-v3/c/include/nvtx3 (same repo, main branch).

Reply:
> The intention was only to disable it for librmm.so

To express that intention, one would use `PRIVATE` rather than `PUBLIC` for `NVTX_DISABLE`. However, for header-only libraries, `PRIVATE` doesn't do anything, since they don't get compiled into a library. Even with the current header-mostly approach, `PRIVATE` wouldn't do much, as it only affects what actually gets compiled (ie, cnmem).

It's all kinda hypothetical at this point, since (a) no nvtx is being used yet and (b) few if any downstream cmake projects use rmm in a way that they'd inherit the flags, or really anything from the cmake targets etc. (Though I guess there's at least one use case of using rmm via `add_subdirectory()` in a downstream project, which does hence have access to the targets.)

Reply:
>  The header only nvtx3 C API has been shipping with the CUDA toolkit at least since CUDA 10. It's the same code that is present here: https://github.com/NVIDIA/NVTX/tree/release-v3/c/include/nvtx3 (same repo, main branch).

Thanks, I see it now in my cuda installation. I agree that `CUDA::nvToolsExt` is not the right target to use when using `nvtx3/*`. One could argue that it'd be nice to have a `CUDA::nvToolsExt3`, which would provide the proper include path, but in practice it'll be found anyway, because we already depend on `CUDA::cudart`, which adds that same include path. If you ever envision a case where you want to use a version of nvtx3 that is not the one that's included with the cuda toolkit in use, it gets iffy, though (kinda like the situation with thrust).

Reply:
> If you ever envision a case where you want to use a version of nvtx3 that is not the one that's included with the cuda toolkit in use, it gets iffy, though (kinda like the situation with thrust).

In the future we'll be depending on the NVTX C++ header only wrapper library here: https://github.com/NVIDIA/NVTX/tree/dev/cpp

We can create a Cmake target for this, but I don't know how to do that, so I'd appreciate help with that.

Reply:
I'll be happy to help. I suppose you'd want to follow the way most dependencies are currently handled, ie., `FetchContent`? That should be fairly straightforward.

Reply:
> I'll be happy to help. I suppose you'd want to follow the way most dependencies are currently handled, ie., `FetchContent`? That should be fairly straightforward.

I think we're moving towards using CPM: https://github.com/rapidsai/cudf/pull/6095

Reply:
Thanks, that's a useful read. I think moving to a more comprehensive package manager is good, but I also have some concerns, so I'll comment on that thread.